File: r/R/io.R
# Internal io helpers. Port of python/src/venn_diagram_lab/io.py:
# _split_line, _detect_delimiter, _parse_table, _read_text.
# Keep this file's helper signatures stable — Phase 2 (TSV exports) and
# Phase 3 (SVG render) call them via the loaders.

#' @importFrom stats setNames
#' @importFrom utils head
NULL

.TRUTHY <- c("1", "true", "yes")
.FALSY  <- c("0", "false", "no", "")
.DELIMITER_CANDIDATES <- c(",", ";", "\t", " ")

.MIN_SETS <- 2L
.MAX_SETS <- 9L

#' @noRd
.split_line <- function(line, delimiter) {
    # Fast path: lines with no double-quote need no stateful quote tracking,
    # so a single vectorised strsplit + trimws reproduces the char-by-char
    # parser below exactly. strsplit() drops trailing empty fields, whereas
    # the reference parser always emits one field per (unquoted) delimiter
    # plus one, so we re-pad to (#delimiters + 1) before trimming.
    if (!grepl('"', line, fixed = TRUE)) {
        n_delim <- nchar(line) - nchar(gsub(delimiter, "", line, fixed = TRUE))
        parts <- strsplit(line, delimiter, fixed = TRUE)[[1L]]
        expected <- n_delim + 1L
        if (length(parts) < expected)
            parts <- c(parts, rep("", expected - length(parts)))
        return(trimws(parts))
    }

    # Slow path: quoted fields may contain the delimiter or escaped quotes.
    result <- character()
    current <- character()
    in_quotes <- FALSE
    chars <- strsplit(line, "", fixed = TRUE)[[1L]]
    n <- length(chars)
    i <- 1L
    while (i <= n) {
        ch <- chars[i]
        if (ch == '"') {
            # Escaped quote inside quoted field?
            if (in_quotes && i + 1L <= n && chars[i + 1L] == '"') {
                current <- c(current, '"')
                i <- i + 2L
                next
            }
            in_quotes <- !in_quotes
        } else if (ch == delimiter && !in_quotes) {
            result <- c(result, trimws(paste(current, collapse = "")))
            current <- character()
        } else {
            current <- c(current, ch)
        }
        i <- i + 1L
    }
    result <- c(result, trimws(paste(current, collapse = "")))
    result
}

#' Pad/truncate each parsed row to a fixed column count, as a character matrix.
#'
#' Mirrors the reference loop's `if (col_idx <= length(row)) row[col_idx] else ""`
#' rule: short rows are right-padded with "", long rows are truncated. Row order
#' is preserved so first-appearance ordering downstream stays byte-identical.
#' @noRd
.rows_to_matrix <- function(rows, ncol) {
    n <- length(rows)
    mat <- matrix("", nrow = n, ncol = ncol)
    for (i in seq_len(n)) {
        r <- rows[[i]]
        m <- min(length(r), ncol)
        if (m > 0L) mat[i, seq_len(m)] <- r[seq_len(m)]
    }
    mat
}

#' @noRd
.detect_delimiter <- function(text) {
    text <- gsub("\r\n", "\n", text, fixed = TRUE)
    text <- gsub("\r", "\n", text, fixed = TRUE)
    lines <- strsplit(trimws(text), "\n", fixed = TRUE)[[1L]]
    lines <- head(lines, 5L)
    if (length(lines) == 0L) return(",")

    best_delim <- ","
    best_score <- -1L
    for (d in .DELIMITER_CANDIDATES) {
        counts <- integer(length(lines))
        for (li in seq_along(lines)) {
            count <- 0L
            in_quotes <- FALSE
            chars <- strsplit(lines[li], "", fixed = TRUE)[[1L]]
            for (ch in chars) {
                if (ch == '"') {
                    in_quotes <- !in_quotes
                } else if (ch == d && !in_quotes) {
                    count <- count + 1L
                }
            }
            counts[li] <- count
        }
        lo <- min(counts); hi <- max(counts)
        if (lo >= 1L && hi - lo <= 1L) {
            score <- lo * 10L + (if (hi == lo) 5L else 0L)
            if (score > best_score) {
                best_score <- score
                best_delim <- d
            }
        }
    }
    best_delim
}

#' @noRd
.read_text <- function(path) {
    text <- paste(readLines(path, warn = FALSE, encoding = "UTF-8"), collapse = "\n")
    list(text = text, source = normalizePath(path, mustWork = TRUE))
}

#' @noRd
.parse_table <- function(text, delimiter) {
    text <- gsub("\r\n", "\n", text, fixed = TRUE)
    text <- gsub("\r", "\n", text, fixed = TRUE)
    raw_lines <- strsplit(trimws(text), "\n", fixed = TRUE)[[1L]]
    lines <- raw_lines[nzchar(trimws(raw_lines))]
    if (length(lines) < 2L)
        .stop_invalid_dataset("File must have at least a header and one data row")
    headers <- .split_line(lines[1L], delimiter)
    rows <- lapply(lines[-1L], .split_line, delimiter = delimiter)
    list(headers = headers, rows = rows)
}

#' @noRd
.binary_columns_to_dataset <- function(headers, rows, source_path, fmt, prefix_cols = 1L) {
    if (length(headers) < prefix_cols + .MIN_SETS)
        .stop_invalid_dataset(
            sprintf("Binary file must have at least %d data columns", .MIN_SETS)
        )

    set_names <- headers[(prefix_cols + 1L):length(headers)]
    n_sets <- length(set_names)
    n_cols <- length(headers)

    # Vectorised reimplementation of the original row-major loop. The cells are
    # already trimmed by .split_line; building a character matrix lets us trim,
    # lower-case and membership-test every cell with a handful of C-level calls
    # instead of growing vectors row by row (which was quadratic on large files).
    mat <- .rows_to_matrix(rows, n_cols)
    item_ids <- trimws(mat[, 1L])
    valid_idx <- which(nzchar(item_ids))
    nonempty_row_count <- length(valid_idx)
    item_order_seen <- unique(item_ids[valid_idx])

    set_cells <- mat[, (prefix_cols + 1L):n_cols, drop = FALSE]
    lc <- set_cells
    lc[] <- tolower(trimws(set_cells))           # preserves matrix dims
    vcells <- lc[valid_idx, , drop = FALSE]      # valid rows x sets

    truthy_mat <- matrix(vcells %in% .TRUTHY, nrow = nonempty_row_count)
    falsy_mat  <- matrix(vcells %in% .FALSY,  nrow = nonempty_row_count)
    bad_mat <- !(truthy_mat | falsy_mat)
    if (any(bad_mat)) {
        # Reconstruct the first invalid cell in the original row-major
        # (row-outer, set-inner) scan order to preserve the error message.
        flat <- which(t(bad_mat))[1L]            # t() => scan sets within a row
        set_index <- ((flat - 1L) %% n_sets) + 1L
        vr_index  <- ((flat - 1L) %/% n_sets) + 1L
        orig_row  <- valid_idx[vr_index]
        raw <- set_cells[orig_row, set_index]
        .stop_invalid_dataset(
            sprintf("Column '%s' row %d has invalid value '%s' (expected 0/1/true/false/yes/no)",
                    set_names[set_index], orig_row + 1L, raw)
        )
    }

    valid_ids <- item_ids[valid_idx]
    items <- setNames(
        lapply(seq_len(n_sets), function(j) unique(valid_ids[truthy_mat[, j]])),
        set_names
    )

    methods::new("VennDataset",
        set_names     = set_names,
        items         = items,
        item_order    = item_order_seen,
        universe_size = nonempty_row_count,
        source_path   = source_path,
        format        = fmt
    )
}

#' @noRd
.aggregated_columns_to_dataset <- function(headers, rows, source_path, fmt) {
    if (length(headers) < .MIN_SETS)
        .stop_invalid_dataset(
            sprintf("Aggregated file must have at least %d columns", .MIN_SETS)
        )

    set_names <- headers
    n_sets <- length(set_names)

    # Vectorised counterpart of the original nested loop (see
    # .binary_columns_to_dataset for the same matrix strategy). Per-set items
    # keep row order then dedupe; `seen` is the first-appearance order across a
    # row-major (row-outer, column-inner) traversal of the non-empty cells.
    mat <- .rows_to_matrix(rows, n_sets)
    mat[] <- trimws(mat)
    items <- setNames(
        lapply(seq_len(n_sets), function(j) {
            col <- mat[, j]
            unique(col[nzchar(col)])
        }),
        set_names
    )
    row_major <- as.vector(t(mat))
    seen <- unique(row_major[nzchar(row_major)])

    if (!any(vapply(items, length, integer(1L)) > 0L))
        .stop_invalid_dataset("Aggregated file has no non-empty cells")

    methods::new("VennDataset",
        set_names     = set_names,
        items         = items,
        item_order    = seen,
        universe_size = NULL,
        source_path   = source_path,
        format        = fmt
    )
}

#' Load a delimited file (CSV/TSV) into a [`VennDataset-class`]
#'
#' Supports two layouts:
#' * Binary mode (default): one row per item, with 0/1 columns marking
#'   membership in each set. The first `prefix_cols` columns are item
#'   metadata; remaining columns are sets.
#' * Aggregated mode (`binary = FALSE`): each column is a set, and cells
#'   contain item identifiers. Empty cells are ignored.
#'
#' @param path Path to the file.
#' @param binary `TRUE` for binary 0/1 mode (default), `FALSE` for aggregated.
#' @param delimiter Explicit delimiter override. `NULL` auto-detects from
#'   `,`, `;`, tab, and space.
#' @param prefix_cols Number of leading metadata columns in binary mode
#'   (default 1). Ignored when `binary = FALSE`.
#' @return A [`VennDataset-class`].
#' @export
#' @examples
#' tmp <- tempfile(fileext = ".csv")
#' writeLines(c("Gene,SetA,SetB", "G1,1,0", "G2,1,1", "G3,0,1"), tmp)
#' ds <- load_csv(tmp, binary = TRUE)
#' ds@set_names
load_csv <- function(path, binary = TRUE, delimiter = NULL, prefix_cols = 1L) {
    txt_src <- .read_text(path)
    delim <- if (is.null(delimiter)) .detect_delimiter(txt_src$text) else delimiter
    parsed <- .parse_table(txt_src$text, delim)
    fmt <- if (delim == "\t") "tsv" else "csv"
    if (isTRUE(binary)) {
        .binary_columns_to_dataset(parsed$headers, parsed$rows, txt_src$source, fmt,
                                    prefix_cols = as.integer(prefix_cols))
    } else {
        .aggregated_columns_to_dataset(parsed$headers, parsed$rows, txt_src$source, fmt)
    }
}

#' Load a tab-separated file into a [`VennDataset-class`]
#'
#' Equivalent to `load_csv(path, binary = binary, delimiter = "\\t", prefix_cols = prefix_cols)`.
#'
#' @inheritParams load_csv
#' @return A [`VennDataset-class`].
#' @export
#' @examples
#' tmp <- tempfile(fileext = ".tsv")
#' writeLines(c("Gene\tSetA\tSetB", "G1\t1\t0", "G2\t1\t1", "G3\t0\t1"), tmp)
#' ds <- load_tsv(tmp, binary = TRUE)
#' ds@universe_size
load_tsv <- function(path, binary = TRUE, prefix_cols = 1L) {
    load_csv(path, binary = binary, delimiter = "\t", prefix_cols = prefix_cols)
}

#' Load a GMT (Gene Matrix Transposed) file into a [`VennDataset-class`]
#'
#' Each line is one set: `set_name<TAB>description<TAB>item1<TAB>item2<TAB>...`.
#' Lines with fewer than 3 tab-separated columns or empty set names are skipped.
#'
#' @param path Path to the .gmt file.
#' @return A [`VennDataset-class`].
#' @export
#' @examples
#' tmp <- tempfile(fileext = ".gmt")
#' writeLines(c("SetA\tdesc\tGENE1\tGENE2\tGENE3",
#'              "SetB\tdesc\tGENE2\tGENE3\tGENE4"), tmp)
#' ds <- load_gmt(tmp)
#' ds@set_names
load_gmt <- function(path) {
    txt_src <- .read_text(path)
    text <- gsub("\r\n", "\n", txt_src$text, fixed = TRUE)
    text <- gsub("\r", "\n", text, fixed = TRUE)
    raw_lines <- strsplit(trimws(text), "\n", fixed = TRUE)[[1L]]
    lines <- raw_lines[nzchar(trimws(raw_lines))]
    if (length(lines) == 0L) .stop_invalid_dataset("GMT file is empty")

    set_names <- character()
    items <- list()
    seen <- character()
    for (line in lines) {
        parts <- strsplit(line, "\t", fixed = TRUE)[[1L]]
        if (length(parts) < 3L) next
        name <- trimws(parts[1L])
        if (!nzchar(name)) next
        members <- trimws(parts[3L:length(parts)])
        members <- members[nzchar(members)]
        if (length(members) == 0L) next
        set_names <- c(set_names, name)
        items[[name]] <- unique(members)
        for (m in members) {
            if (!m %in% seen) seen <- c(seen, m)
        }
    }

    if (length(set_names) == 0L)
        .stop_invalid_dataset("GMT file has no valid gene sets")
    if (length(set_names) < .MIN_SETS)
        .stop_invalid_dataset(
            sprintf("GMT file must contain at least %d sets, got %d",
                    .MIN_SETS, length(set_names))
        )
    if (length(set_names) > .MAX_SETS)
        .stop_invalid_dataset(
            sprintf("GMT file has %d sets; max supported is %d. Filter the file before loading.",
                    length(set_names), .MAX_SETS)
        )

    methods::new("VennDataset",
        set_names     = set_names,
        items         = items,
        item_order    = seen,
        universe_size = NULL,
        source_path   = txt_src$source,
        format        = "gmt"
    )
}

#' Load a GMX file (transposed GMT) into a [`VennDataset-class`]
#'
#' Row 0 = set names, row 1 = descriptions, rows 2+ = items column-aligned.
#'
#' @param path Path to the .gmx file.
#' @return A [`VennDataset-class`].
#' @export
#' @examples
#' tmp <- tempfile(fileext = ".gmx")
#' writeLines(c("SetA\tSetB",
#'              "desc_A\tdesc_B",
#'              "GENE1\tGENE2",
#'              "GENE2\tGENE3"), tmp)
#' ds <- load_gmx(tmp)
#' length(ds@items)
load_gmx <- function(path) {
    txt_src <- .read_text(path)
    text <- gsub("\r\n", "\n", txt_src$text, fixed = TRUE)
    text <- gsub("\r", "\n", text, fixed = TRUE)
    raw_lines <- strsplit(trimws(text), "\n", fixed = TRUE)[[1L]]
    lines <- raw_lines[nzchar(trimws(raw_lines))]
    if (length(lines) < 3L)
        .stop_invalid_dataset("GMX file must have at least 3 rows (names, descriptions, genes)")

    header_parts <- strsplit(lines[1L], "\t", fixed = TRUE)[[1L]]
    set_names <- trimws(header_parts)
    set_names <- set_names[nzchar(set_names)]
    if (length(set_names) < .MIN_SETS)
        .stop_invalid_dataset(
            sprintf("GMX file must have at least %d columns", .MIN_SETS)
        )
    if (length(set_names) > .MAX_SETS)
        .stop_invalid_dataset(
            sprintf("GMX file has %d sets; max supported is %d.",
                    length(set_names), .MAX_SETS)
        )

    items <- setNames(replicate(length(set_names), character(0), simplify = FALSE), set_names)
    seen <- character()
    for (line in lines[3L:length(lines)]) {
        parts <- trimws(strsplit(line, "\t", fixed = TRUE)[[1L]])
        for (col_idx in seq_along(set_names)) {
            cell <- if (col_idx <= length(parts)) parts[col_idx] else ""
            if (nzchar(cell)) {
                items[[set_names[col_idx]]] <- c(items[[set_names[col_idx]]], cell)
                if (!cell %in% seen) seen <- c(seen, cell)
            }
        }
    }
    items <- lapply(items, unique)

    methods::new("VennDataset",
        set_names     = set_names,
        items         = items,
        item_order    = seen,
        universe_size = NULL,
        source_path   = txt_src$source,
        format        = "gmx"
    )
}

File: python/src/venn_diagram_lab/io.py
"""Data-loading and parsing for venn-diagram-lab.

Supports CSV / TSV (binary or aggregated mode), GMT, GMX. Always returns a
canonical `Dataset` so downstream modules don't care about the source format.
"""

from __future__ import annotations

from collections.abc import Iterable, Mapping
from dataclasses import dataclass
from pathlib import Path
from typing import Literal

from venn_diagram_lab.errors import InvalidDatasetError

DatasetFormat = Literal["csv", "tsv", "gmt", "gmx"]

MIN_SETS = 2
MAX_SETS = 9


@dataclass(frozen=True)
class Dataset:
    """In-memory representation of a Venn-diagram input.

    Attributes:
        set_names: Ordered set identifiers (preserving the order they
            appeared in the source).
        items: Map ``set_name -> set of item identifiers`` (gene symbols,
            titles, etc.).
        source_path: Original file path if loaded from disk; ``None`` for
            in-memory or sample-derived datasets.
        format: Source format (``csv``/``tsv``/``gmt``/``gmx``). For in-memory
            datasets the default is ``csv``; callers that care about format
            may override.
        item_order: First-seen insertion order across all sets, mirroring JS
            ``Set``/``Map`` semantics. Empty tuple when not populated (legacy
            callers). All bundled loaders and ``from_dict`` populate it.
        universe_size: Hypergeometric universe N from the source file, when
            known. Binary CSV/TSV loaders set this to the row count
            (matching the React webapp's ``csv.rows.length``); other formats
            leave it ``None``, signalling "compute as |union of items|"
            downstream.
    """

    set_names: list[str]
    items: dict[str, set[str]]
    source_path: str | None
    format: DatasetFormat
    item_order: tuple[str, ...] = ()
    universe_size: int | None = None

    @classmethod
    def from_dict(
        cls,
        data: Mapping[str, Iterable[str]],
        *,
        source_path: str | None = None,
        format: DatasetFormat = "csv",
    ) -> Dataset:
        """Build a Dataset from a plain dict of set_name -> iterable of items.

        Items are coerced to `str` and de-duplicated. Set count must be in [2, 9].
        Iterables are consumed exactly once; generators are safe to pass.
        """
        if len(data) < MIN_SETS:
            raise InvalidDatasetError(
                f"Dataset must have at least {MIN_SETS} sets, got {len(data)}"
            )
        if len(data) > MAX_SETS:
            raise InvalidDatasetError(
                f"Dataset must have at most {MAX_SETS} sets, got {len(data)}"
            )
        names = list(data.keys())
        # Materialise each iterable to a list so we can scan it twice (item_order + items).
        materialised: dict[str, list[str]] = {
            name: [str(x) for x in values] for name, values in data.items()
        }
        items = {name: set(vals) for name, vals in materialised.items()}
        seen: dict[str, None] = {}
        for name in names:
            for v in materialised[name]:
                if v not in seen:
                    seen[v] = None
        return cls(
            set_names=names,
            items=items,
            source_path=source_path,
            format=format,
            item_order=tuple(seen.keys()),
        )


def _split_line(line: str, delimiter: str) -> list[str]:
    """Split a CSV line respecting quoted fields and escaped quotes.

    Mirrors src/utils/csvParser.ts splitCsvLineWithDelimiter exactly so loaded
    datasets parse identically to the web tool.
    """
    result: list[str] = []
    current: list[str] = []
    in_quotes = False
    i = 0
    while i < len(line):
        ch = line[i]
        if ch == '"':
            if in_quotes and i + 1 < len(line) and line[i + 1] == '"':
                current.append('"')
                i += 2
                continue
            in_quotes = not in_quotes
        elif ch == delimiter and not in_quotes:
            result.append("".join(current).strip())
            current = []
        else:
            current.append(ch)
        i += 1
    result.append("".join(current).strip())
    return result


_TRUTHY = {"1", "true", "yes"}
_FALSY = {"0", "false", "no", ""}

_DELIMITER_CANDIDATES: tuple[str, ...] = (",", ";", "\t", " ")


def _detect_delimiter(text: str) -> str:
    """Auto-detect the delimiter by scoring consistency across the first 5 lines.

    Mirrors src/utils/csvParser.ts detectDelimiter. Falls back to ',' if no
    candidate scores above zero (single-token lines, etc.).
    """
    lines = text.replace("\r\n", "\n").replace("\r", "\n").strip().split("\n")[:5]
    if not lines:
        return ","

    best_delim = ","
    best_score = -1
    for d in _DELIMITER_CANDIDATES:
        counts = []
        for line in lines:
            count = 0
            in_quotes = False
            for ch in line:
                if ch == '"':
                    in_quotes = not in_quotes
                elif ch == d and not in_quotes:
                    count += 1
            counts.append(count)
        if not counts:
            continue
        lo, hi = min(counts), max(counts)
        if lo >= 1 and hi - lo <= 1:
            score = lo * 10 + (5 if hi == lo else 0)
            if score > best_score:
                best_score = score
                best_delim = d
    return best_delim


# ---------------------------------------------------------------------------
# File I/O helpers
# ---------------------------------------------------------------------------


def _read_text(path: Path | str) -> tuple[str, str]:
    """Read the file contents and return (text, str(path))."""
    p = Path(path)
    return p.read_text(encoding="utf-8"), str(p)


def _parse_table(text: str, delimiter: str) -> tuple[list[str], list[list[str]]]:
    """Low-level: parse a delimited file into (headers, rows)."""
    lines = [
        line
        for line in text.replace("\r\n", "\n").replace("\r", "\n").strip().split("\n")
        if line.strip()
    ]
    min_lines = 2  # 1 header + at least 1 data row
    if len(lines) < min_lines:
        raise InvalidDatasetError("File must have at least a header and one data row")
    headers = _split_line(lines[0], delimiter)
    rows = [_split_line(line, delimiter) for line in lines[1:]]
    return headers, rows


def _binary_columns_to_dataset(
    headers: list[str],
    rows: list[list[str]],
    source_path: str,
    fmt: DatasetFormat,
    prefix_cols: int = 1,
) -> Dataset:
    """Treat the first *prefix_cols* columns as item metadata, remaining as binary sets.

    Parameters
    ----------
    prefix_cols:
        Number of leading columns to treat as item metadata (default 1).
        Column 0 is always the item id used for set membership; columns
        1..(prefix_cols-1) are ignored metadata fields.
    """
    if len(headers) < prefix_cols + MIN_SETS:
        raise InvalidDatasetError(
            f"Binary file must have at least {MIN_SETS} data columns"
        )

    set_names = headers[prefix_cols:]
    items: dict[str, set[str]] = {name: set() for name in set_names}
    item_order_seen: dict[str, None] = {}
    nonempty_row_count = 0

    for row_idx, row in enumerate(rows, start=2):  # row 1 is the header
        if not row or not row[0].strip():
            continue
        nonempty_row_count += 1
        item_id = row[0].strip()
        if item_id not in item_order_seen:
            item_order_seen[item_id] = None
        for col_offset, set_name in enumerate(set_names):
            col_idx = prefix_cols + col_offset
            cell = (row[col_idx] if col_idx < len(row) else "").strip().lower()
            if cell in _TRUTHY:
                items[set_name].add(item_id)
            elif cell in _FALSY:
                continue
            else:
                raw = row[col_idx] if col_idx < len(row) else ""
                raise InvalidDatasetError(
                    f"Column {set_name!r} row {row_idx} has invalid value {raw!r} "
                    "(expected 0/1/true/false/yes/no)"
                )

    return Dataset(
        set_names=set_names,
        items=items,
        source_path=source_path,
        format=fmt,
        item_order=tuple(item_order_seen.keys()),
        universe_size=nonempty_row_count,
    )


def _aggregated_columns_to_dataset(
    headers: list[str],
    rows: list[list[str]],
    source_path: str,
    format: DatasetFormat,
) -> Dataset:
    """Treat each column as a set; non-empty cells are item identifiers."""
    if len(headers) < MIN_SETS:
        raise InvalidDatasetError(f"Aggregated file must have at least {MIN_SETS} columns")

    set_names = list(headers)
    items: dict[str, set[str]] = {name: set() for name in set_names}
    seen: dict[str, None] = {}

    for row in rows:
        for col_idx, set_name in enumerate(set_names):
            cell = (row[col_idx] if col_idx < len(row) else "").strip()
            if cell:
                items[set_name].add(cell)
                if cell not in seen:
                    seen[cell] = None

    if not any(items.values()):
        raise InvalidDatasetError("Aggregated file has no non-empty cells")

    return Dataset(
        set_names=set_names,
        items=items,
        source_path=source_path,
        format=format,
        item_order=tuple(seen.keys()),
    )


# ---------------------------------------------------------------------------
# Public loaders
# ---------------------------------------------------------------------------


def load_csv(
    path: Path | str,
    *,
    binary: bool = True,
    delimiter: str | None = None,
    prefix_cols: int = 1,
) -> Dataset:
    """Load a delimited file into a :class:`Dataset`.

    Auto-detects the delimiter from ``,``, ``;``, ``<tab>``, and ``<space>``
    if not given. Supports two layouts:

    * **Binary mode** (default): one row per item, with 0/1 columns marking
      membership in each set. The first column (or first ``prefix_cols``
      columns) is item metadata; remaining columns are sets.
    * **Aggregated mode** (``binary=False``): each column is a set, and cells
      contain item identifiers. Empty cells are ignored.

    Args:
        path: Path to the file (str or Path).
        binary: Treat the file as binary 0/1 (default) or aggregated.
        delimiter: Explicit delimiter. ``None`` auto-detects.
        prefix_cols: Number of leading metadata columns in binary mode
            (default 1, meaning column 0 is the item id). Ignored when
            ``binary=False``.

    Returns:
        Dataset with ``set_names``, ``items``, and ``item_order`` populated.
        ``universe_size`` is the row count for binary files; ``None`` otherwise.

    Example:
        >>> from venn_diagram_lab import load_csv
        >>> ds = load_csv("genes.csv", binary=True)
        >>> ds.set_names
        ['SetA', 'SetB', 'SetC']
        >>> len(ds.items["SetA"])
        42
    """
    text, src = _read_text(path)
    delim = delimiter if delimiter is not None else _detect_delimiter(text)
    headers, rows = _parse_table(text, delim)
    fmt: DatasetFormat = "tsv" if delim == "\t" else "csv"
    if binary:
        return _binary_columns_to_dataset(headers, rows, src, fmt, prefix_cols=prefix_cols)
    return _aggregated_columns_to_dataset(headers, rows, src, fmt)


def load_tsv(path: Path | str, *, binary: bool = True, prefix_cols: int = 1) -> Dataset:
    """Load a tab-separated file into a :class:`Dataset`.

    Equivalent to ``load_csv(path, binary=binary, delimiter='\\t',
    prefix_cols=prefix_cols)``.

    Args:
        path: Path to the .tsv file.
        binary: Binary 0/1 mode (default) or aggregated (cells = item names).
        prefix_cols: Leading metadata columns in binary mode (default 1).

    Returns:
        Dataset.

    Example:
        >>> from venn_diagram_lab import load_tsv
        >>> ds = load_tsv("cancer_drivers.tsv", binary=True)
        >>> ds.universe_size  # source row count
        20000
    """
    return load_csv(path, binary=binary, delimiter="\t", prefix_cols=prefix_cols)


def load_gmt(path: Path | str) -> Dataset:
    """Load a GMT (Gene Matrix Transposed) file into a :class:`Dataset`.

    Each line is one set: ``set_name<TAB>description<TAB>item1<TAB>item2<TAB>...``.

    Args:
        path: Path to the .gmt file.

    Returns:
        Dataset.

    Example:
        >>> from venn_diagram_lab import load_gmt
        >>> ds = load_gmt("hallmark.gmt")
        >>> ds.set_names[:3]
        ['HALLMARK_HYPOXIA', 'HALLMARK_APOPTOSIS', 'HALLMARK_INFLAMMATORY_RESPONSE']
    """
    text, src = _read_text(path)
    lines = [
        line
        for line in text.replace("\r\n", "\n").replace("\r", "\n").strip().split("\n")
        if line.strip()
    ]
    if not lines:
        raise InvalidDatasetError("GMT file is empty")

    set_names: list[str] = []
    items: dict[str, set[str]] = {}
    seen: dict[str, None] = {}

    gmt_min_cols = 3  # name, description, at least one gene
    for line in lines:
        parts = line.split("\t")
        if len(parts) < gmt_min_cols:
            continue  # need at least name, description, one gene
        name = parts[0].strip()
        if not name:
            continue
        members = [p.strip() for p in parts[2:] if p.strip()]
        if not members:
            continue
        set_names.append(name)
        items[name] = set(members)
        for m in members:
            if m not in seen:
                seen[m] = None

    if not set_names:
        raise InvalidDatasetError("GMT file has no valid gene sets")
    if len(set_names) < MIN_SETS:
        raise InvalidDatasetError(
            f"GMT file must contain at least {MIN_SETS} sets, got {len(set_names)}"
        )
    if len(set_names) > MAX_SETS:
        raise InvalidDatasetError(
            f"GMT file has {len(set_names)} sets; the current version supports "
            f"at most {MAX_SETS} (limited by the bundled Venn templates, which "
            "cover 2-9 sets). Filter the file to <= 9 sets before loading. "
            "UpSet rendering for >9 sets without going through analyze() is on "
            "the roadmap for a future release; track it at "
            "https://github.com/ZoliQua/Venn-Diagram-Lab/issues."
        )

    return Dataset(
        set_names=set_names,
        items=items,
        source_path=src,
        format="gmt",
        item_order=tuple(seen.keys()),
    )


def load_gmx(path: Path | str) -> Dataset:
    """Load a GMX file (transposed GMT) into a :class:`Dataset`.

    Row 0 = set names, row 1 = descriptions, rows 2+ = items column-aligned.

    Args:
        path: Path to the .gmx file.

    Returns:
        Dataset.

    Example:
        >>> from venn_diagram_lab import load_gmx
        >>> ds = load_gmx("pathways.gmx")
        >>> sum(len(items) for items in ds.items.values())
        1234
    """
    text, src = _read_text(path)
    lines = [
        line
        for line in text.replace("\r\n", "\n").replace("\r", "\n").strip().split("\n")
        if line.strip()
    ]
    min_rows = 3
    if len(lines) < min_rows:
        raise InvalidDatasetError("GMX file must have at least 3 rows (names, descriptions, genes)")

    set_names = [h.strip() for h in lines[0].split("\t") if h.strip()]
    if len(set_names) < MIN_SETS:
        raise InvalidDatasetError(f"GMX file must have at least {MIN_SETS} columns")
    if len(set_names) > MAX_SETS:
        raise InvalidDatasetError(
            f"GMX file has {len(set_names)} sets; the current version supports "
            f"at most {MAX_SETS} (limited by the bundled Venn templates, which "
            "cover 2-9 sets). Filter the file to <= 9 sets before loading. "
            "UpSet rendering for >9 sets without going through analyze() is on "
            "the roadmap for a future release; track it at "
            "https://github.com/ZoliQua/Venn-Diagram-Lab/issues."
        )

    items: dict[str, set[str]] = {name: set() for name in set_names}
    seen: dict[str, None] = {}
    # Row-major scan: top-to-bottom rows, left-to-right columns within each row.
    # Matches the webapp's Set insertion order when GMX is read top-to-bottom.
    for line in lines[2:]:
        parts = [p.strip() for p in line.split("\t")]
        for col_idx, set_name in enumerate(set_names):
            cell = parts[col_idx] if col_idx < len(parts) else ""
            if cell:
                items[set_name].add(cell)
                if cell not in seen:
                    seen[cell] = None

    return Dataset(
        set_names=set_names,
        items=items,
        source_path=src,
        format="gmx",
        item_order=tuple(seen.keys()),
    )
